src/Orleans/Messaging/ProxiedMessageCenter.cs

```diff
@@ -287,6 +293,12 @@ public Message WaitMessage(Message.Categories type, CancellationToken ct)
         {
             try
             {
+                if (ct.IsCancellationRequested)
+                {
+                    return null;
+                }
+
+                // Don't pass CancellationToken to Take. It causes too much spinning.
                 Message msg = PendingInboundMessages.Take();
```

Comment:
Regarding the comment at https://github.com/dotnet/orleans/issues/1350#issuecomment-176641395 :

hmmm, that's very odd, but I guess it has to do with Disposing the blocking collection at the same time that the `Take` is initiating, and getting into a weird state. I was trying to follow the code in the framework, but couldn't find an `ArgumentNullException` being thrown from `SemaphoreSlim.Wait` directly. Probably some function got inlined and it's not showing in the stack trace, and the code is very tough to follow. Any chance we can delay disposing somehow? If not, it might make to catch ArgumentNullException (although I dislike that, but it's better than having a catch-all clause for all exceptions to be treated equally, when we know this exception happens harmlessly)